Don't announce 'selected' when the focus moves in Google sheets if the focused cell is the only cell selected

michaelDCurran opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
Fixes #8766 

### Summary of the issue:
When navigating cells in Google Sheets with GoogleShets Braille mode enabled, NVDA reports 'selected' for every cell the focus moves to.
Although this is technically correct (the cell is selected), it is not at all practical. Most if not all cells that are focused will be selected. So it ends up being redundant information.
We should change the experience so that:
* If the focused cell is the only cell selected, then NvDA should not announce that it is selected.
* If the focused cell is selected and more than one cell is selected, NVDA should announce that the focused cell is selected.
* If the focused cell is not selected, NVDA should announce that the focused cell is 'not selected'.

### Description of how this pull request fixes the issue:
This PR implements these rules by the following:
* NvDAObjects now have a selectionContainer property, which should return the ancestor NVDAObject that manages the selection. E.g. for a table cell it would be the table. For a list item it would be the list. By default this property returns None.
* Implement the selectionContainer property on IAccessible NVDAObjects. This implementation currently just returns the table NVDAObject if the NVDAObject has an ancestor table. In the future, other NVDAObjects such as UIA should implement this to make use of the UIA selection pattern etc.
* Implement a getSelectedItemsCount method on NVDAObjects. This method should return the number of descendants currently selected. For performance reasons, the method takes a maxCount argument, which can be used to limet the amount of items counted, if an implementation has to literally fetch the items using a preallocated buffer in order to get the number of them. Implementations should try and fetch one more than maxCount, and if there are more items than maxCount, then sys.maxint should be returned, to denote many items.
* Implemented getSelectedItemsCount on IAccessible NVDAObject. This implementation tries to fetch and count items with IAccessible::accSelect, and if that fails or is not available, then if the IAccessible supports IAccessibleTable2, IAccessibleTable2::nSelectedCells is used instead. Note that As Chrome does not implement accSelection on tables, it is necessary to fall back to nSelectedCells.
* speech.speakObjectProperties: When speaking object properties for the reason of focus, remove the selected and selectable states from the object's states to be spoken, if the selectable state is present, the object has a selectionContainer, and its selectionContainer's getSelectedItemsCount is equal to 1.
 * controlTypes.processNegativeStates:  report 'not selected' when there is the selectable state but not the selected state, if the role is tableCell, tableColumnHeader or tableRowHeader. Previously we were already doing this for tableRow.

 
 
### Testing performed:
Created a new sheet in Google Sheets. Moved focus around the cells. NVDA no longer announced 'selected' if the focused cell was the only cell selected.

### Known issues with pull request:
None.

### Change log entry:
Bug fixes:
* In Google Sheets with braille mode enabled, NVDA no longer announces 'selected' on every cell when moving focus between cells.

michaelDCurran committed:
Don't announce 'selected' when the focus moves in Google sheets if the focused cell is the only cell selected.

leonardder commented:
I updated the initial description in order for it to automatically close #8766

leonardder commented on source/NVDAObjects/IAccessible/__init__.py:
```diff
@@ -1214,6 +1216,38 @@ def _get_rowHeaderText(self):
 
+	def _get_selectionContainer(self):
+		if self.table:
+			return self.table
```
This implicitly returns None when not self.table. I would prefer an explicit return of None or even better, super below this line, though not strictly necessary. I leave it up to you.

leonardder commented on source/controlTypes.py:
```diff
@@ -727,7 +727,7 @@ def processNegativeStates(role, states, reason, negativeStates=None):
 	# The condition stops "not selected" from being spoken in some broken controls
 	# when the state change for the previous focus is issued before the focus change.
-	if role in (ROLE_LISTITEM, ROLE_TREEVIEWITEM, ROLE_TABLEROW) and STATE_SELECTABLE in states and (reason != REASON_CHANGE or STATE_FOCUSED in states):
+	if role in (ROLE_LISTITEM, ROLE_TREEVIEWITEM, ROLE_TABLEROW,ROLE_TABLECELL,ROLE_TABLECOLUMNHEADER,ROLE_TABLEROWHEADER) and STATE_SELECTABLE in states and (reason != REASON_CHANGE or STATE_FOCUSED in states):
```
Could you split this into multiple lines?

LeonarddeR reviewed (commented):
I'd be curious what this does with braille, but I"m currently unable to test with braille. I'd say we also want to apply the "don't present selected if only one item is selected" case to braille as well?

michaelDCurran committed:
Address review comments.

michaelDCurran commented:
@leonardder: Hi have addressed your review comments for 
IAccessible.getSelectedItemsCount and 
controlTypes.processNegativeStates. However:

In regards to braille: because Braille does not currently show negative 
states such as 'not selected', if we took away 'selected' in this case 
there would be an ambiguity. And this would have always been the case, 
not just for Google Sheets.

My feeling is that we should just leave braille changes out for now, 
until we can come up with a better presentation of 'not selected' verses 
'selected' in braille... unless I'm not understanding you correctly here.

michaelDCurran committed:
Merge branch 'master' into getSelectedItemsCount

michaelDCurran committed:
Address review comments.

michaelDCurran commented:
@feerrenrut I split out the accSelection stuff into its own helper method. But as the IAccessibleTable2 code is only one call I left it in the original method. I also addressed your other review comments.

michaelDCurran committed:
Update what's new

michaelDCurran commented:
This pr is causing 'not selected' to be announced when navigating by table cell in browse mode.
